File: controls/TypedValue.js
/*
 *  Copyright 2013 Juha Lepola
 *
 *  This file is part of M-Files for Sailfish.
 *
 *  M-Files for Sailfish is free software: you can redistribute it
 *  and/or modify it under the terms of the GNU General Public
 *  License as published by the Free Software Foundation, either
 *  version 3 of the License, or (at your option) any later version.
 *
 *  M-Files for Sailfish is distributed in the hope that it will be
 *  useful, but WITHOUT ANY WARRANTY; without even the implied warranty
 *  of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with M-Files for Sailfish. If not, see
 *  <http://www.gnu.org/licenses/>.
 */

.import Sailfish.Silica 1.0 as Silica

var __TVMFDataTypes = {

	Text : 1,
	Integer : 2,
	Floating : 3,
	Date : 5,
	Time : 6,
	Timestamp : 7,
	Boolean : 8,
	Lookup : 9,
	MultiLineText : 13,
	MultiSelectLookup : 10,
}

/**
 * Selects the control for displaying the typed value.
 *
 * @param {typedValue} Typed value
 */
function selectControl( typedValue ) {

	// Select the control based on the data type.
	switch( typedValue.DataType ) {

	// Single-line text
	case 1:
		return "SingleLineText.qml";

	// Integer
	case 2:
		return "Integer.qml";

	// Real
	case 3:
		return "Real.qml";

	// Date
	case 5:
		return "Date.qml";

	// Time
	case 6:
		return "Time.qml";

	// Boolean
	case 8:
		return "Boolean.qml";

	// Single-select lookup
	case 9:
		return "Lookup.qml";

	// Multi-select lookup
	case 10:
		return "Lookups.qml";

	case 13:
		return "MultiLineText.qml";

	// We return simple read-only control for types we do not recognize.
	default:
		return "ReadOnlyLabel.qml";
	}
}

/**
 * Selects implicit height for the typed value control.
 *
 * @param {typedValue} Typed value
 */
function selectImplicitHeight( typedValue ) {

	// Implicit height based on the data type.
	switch( typedValue.DataType ) {

	// Single-line text
	case 1:
		return Silica.Theme.itemSizeExtraSmall;

	// Multi-select lookup
	case 10:
		var lookups = typedValue.Lookups
		return selectImplicitHeightForMSLookup( lookups );

	case 13:
		return Silica.Theme.itemSizeExtraLarge * 2;

	// We return simple read-only control for types we do not recognize.
	default:
		return Silica.Theme.itemSizeExtraSmall;
	}

}


/**
 * Selects implicit height for the typed value control that represents multi-select lookup.
 *
 * @param {lookups} Lookups
 */
function selectImplicitHeightForMSLookup( lookups )
{

	var itemsShown = lookups.length;
	if( itemsShown > 2 )
		itemsShown = 3;
	return itemsShown * Silica.Theme.itemSizeExtraSmall;
}

File: src/mfiles/typedvalue.cpp
/*
 *  Copyright 2014 Juha Lepola
 *
 *  This file is part of M-Files for Sailfish.
 *
 *  M-Files for Sailfish is free software: you can redistribute it
 *  and/or modify it under the terms of the GNU General Public
 *  License as published by the Free Software Foundation, either
 *  version 3 of the License, or (at your option) any later version.
 *
 *  M-Files for Sailfish is distributed in the hope that it will be
 *  useful, but WITHOUT ANY WARRANTY; without even the implied warranty
 *  of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with M-Files for Sailfish. If not, see
 *  <http://www.gnu.org/licenses/>.
 */

#include "typedvalue.h"

#include "mfilesconstants.h"

#include <QDate>
#include <QDateTime>
#include <QDebug>
#include <QJsonArray>
#include <QUrl>

/**
 * Namespace for M-Files types.
 */
namespace MFiles
{

TypedValue::TypedValue( const QJsonValue& typedValue ) :
	MFilesTypeWrapper( typedValue )
{
}

/**
 * @brief Initializes a new lookup based typed value
 */
TypedValue::TypedValue( int dataType, const Lookup& lookup )
{
	qDebug( "TypedValue" );
	Q_ASSERT( dataType == 9 || dataType == 10 );

	// Get reference to the internal typed value variable and set the values.
	QJsonObject& typedValue = this->object();
	typedValue[ "DataType" ] = dataType;
	typedValue[ "HasValue" ] = ! lookup.isUndefined();
	typedValue[ "DisplayValue" ] = lookup.displayValue();

	// Set the value based on the data type.
	switch ( dataType )
	{
	// Single-select.
	case MFiles::Constants::SingleSelectLookup :
	{
		typedValue[ "Lookup" ] = lookup.toJsonValue();
		typedValue[ "Value" ] = lookup.toJsonValue();
		break;
	}

	// Multi-select
	case MFiles::Constants::MultiSelectLookup :
	{
		QJsonArray lookups;
		lookups.push_back( lookup.toJsonValue() );
		typedValue[ "Lookups" ] = lookups;
		typedValue[ "Value" ] = lookups;
		break;
	}

	// Unexpected data type.
	default:
		qCritical( "TODO: Report error." );
		break;
	}
}

TypedValue::TypedValue( const QJsonArray& lookups )
{
	// Delegate.
	this->setMultiSelectLookup( lookups );
}

TypedValue::TypedValue( const TypedValue& source ) :
	MFilesTypeWrapper( source.toJsonValue() )
{

}

/**
 * @brief asLookups
 * @return All lookups as a Json array. Can be called for both single-select and multi-select lookups.
 */
QJsonArray TypedValue::asLookups() const
{
	QJsonArray lookups;
	if( ! this->hasValue() )
		return lookups;
	switch ( this->dataType() )
	{
	// Single-select.
	case 9:
	{
		lookups.push_back( this->object()[ "Lookup" ] );
		break;
	}

	// Multi-select
	case 10:
	{
		foreach( Lookup lookup, this->object()[ "Lookups" ].toArray() )
		{
			lookups.push_back( lookup.toJsonValue() );
		}
		break;
	}

	// Unexpected data type.
	default:
		qCritical( "TODO: Report error." );
		break;
	}

	return lookups;
}

/**
 * @brief getLookupIds
 * @return A collection of lookup ids.
 */
QSet< int > TypedValue::getLookupIds() const
{
	Q_ASSERT( this->dataType() == 9 || this->dataType() == 10 );

	// Collect the ids of the lookups.
	QSet< int > lookupIds;
	foreach( Lookup lookup, this->asLookups() )
	{
		lookupIds.insert( lookup.item() );
	}
	return lookupIds;
}

void TypedValue::setMultiSelectLookup( const QJsonArray& lookups )
{
	// Get reference to the internal typed value variable and set the values.
	QJsonObject& typedValue = this->object();
	typedValue[ "DataType" ] = MFiles::Constants::MultiSelectLookup;
	typedValue[ "HasValue" ] = lookups.size() != 0;
	typedValue[ "DisplayValue" ] = QString( "Demo" );
	typedValue[ "Lookups" ] = lookups;
	typedValue[ "Value" ] = lookups;
	Q_ASSERT( this->hasValue() || lookups.count() == 0 );
	Q_ASSERT( this->asLookups().count() == lookups.count() );
}

bool TypedValue::dropLookupsExcept( const QSet< int > allowedLookups )
{
	// Stop if we do not have a value.
	if( ! this->hasValue() )
		return false;

	bool somethingDropped = false;
	QJsonObject& typedValue = this->object();
	switch ( this->dataType() )
	{
	// Single-select.
	case MFiles::Constants::SingleSelectLookup :
	{
		// Drop the only lookup if not included in allowedLookups.
		Lookup lookup( typedValue[ "Lookup" ] );
		if( ! allowedLookups.contains( lookup.item() ) )
		{
			// The lookup value should be dropped.
			somethingDropped = true;
			typedValue[ "Lookup" ] = QJsonValue::Null;
			typedValue[ "Value" ] = QJsonValue::Null;
			typedValue[ "DisplayValue" ] = QString( "" );
			typedValue[ "HasValue" ] = false;
		}
		break;
	}

	// Multi-select
	case MFiles::Constants::MultiSelectLookup:
	{
		// Get current lookups and drop all those lookup values that are not allowed.
		// Then set them back to this typed value.
		QJsonArray lookups = typedValue[ "Lookups" ].toArray();
		for( QJsonArray::iterator itr = lookups.begin();
			 itr != lookups.end();  )
		{
			// Discard all lookups that are not allowed.
			Lookup lookup( *itr );
			if( allowedLookups.contains( lookup.item() ) )
			{
				// This lookup is allowed.
				itr++;
			}
			else
			{
				// This lookup is not allowed. Erase it.
				somethingDropped = true;
				itr = lookups.erase( itr );
			}

		}
		this->setMultiSelectLookup( lookups );
		break;
	}

	// Unexpected data type.
	default:
		qCritical( "TODO: Report error." );
		break;
	}

	return somethingDropped;
}

QString TypedValue::getUriEncodedValue() const
{
	// Uninitialized?
	if( ! hasValue() )
		return "-";

	// Determine the encoding prefix and the encoded value.
	char prefix;
	QString value = "";
	switch ( this->dataType() )
	{
	case MFiles::Constants::Text :
		prefix = 'T';
		value = QUrl::toPercentEncoding( this->getAsUnlocalizedText() ).constData();
		break;

	case MFiles::Constants::Integer :
		prefix = 'I';
		value = QString::number( this->value().toDouble() );
		break;

	case MFiles::Constants::Floating :
		prefix = 'R';
		value = QString::number( this->value().toDouble() );
		break;

	case MFiles::Constants::Date :
		{
			prefix = 'D';
			QDate date = QDate::fromString( this->value().toString(), Qt::ISODate );
			value = date.toString( "yyyy-MM-dd" );
		}
		break;

	// TODO: Need to investagte how to properly format the time and timestamps fields.
	/*case MFiles::Constants::Time :
		{
			prefix = 'C';
			qDebug() << this->value().toString();
			QDateTime dt = QDateTime::fromString( this->value().toString(), Qt::ISODate );
			value = dt.toString( "yyyy-MM-ddThh:mm:ss" );
		}
		break;

	case MFiles::Constants::Timestamp :
		prefix = 'P';
		value = this->value().toString();
		break;*/

	case MFiles::Constants::SingleSelectLookup :
		prefix = 'L';
		value = QString::number( this->lookup().item() );
		break;

	// Multi-select
	case MFiles::Constants::MultiSelectLookup:
		prefix = 'S';
		foreach( int lookupId, this->getLookupIds() )
		{
			value.append( QString::number( lookupId ) );
			value.append( ',' );
		}
		value.chop( 1 );  // Remove the extra comma.
		break;

	case MFiles::Constants::MultiLineText :
		prefix = 'M';
		value = QUrl::toPercentEncoding( this->getAsUnlocalizedText() ).constEnd();
		break;

	// Unexpected data type.
	default:
		prefix = '-';
		qCritical( "TODO: Report error." );
		break;
	}

	// Format the value.
	QString format( "%1%2" );
	QString encodedValue = format.arg( prefix ).arg( value );

	// Double URI encode.
	encodedValue = QUrl::toPercentEncoding( encodedValue );
	// encodedValue.replace( "%", "_p" );
	return encodedValue;
}

QString TypedValue::getAsUnlocalizedText() const
{
	// Uninitialized?
	if( ! hasValue() )
		return "";

	QString value;
	switch ( this->dataType() )
	{
	case MFiles::Constants::Text :
	case MFiles::Constants::MultiLineText :
		value = this->property( "Value" ).toString();
		break;

	// Unexpected data type.
	default:
		qCritical( "TODO: Report error." );
		break;
	}
	return value;
}

}
